### ISSUE TITLE
rospy: Subscriber memory leak from Statistics circular-ref

### ISSUE THREAD

**Issue by rethink-rlinsalata** [ros/ros_comm#520]
**Issue**: rospy Subscribers cause a memory leak when being garabage collected.

**Cause**: This is due to a circular-reference in the creation of the child member `SubscriberStatisticsLogger(self)` which includes a reference to the parent Subscriber.

**Test**:
This can be tested by looking at the garbage collection after creating a Subscriber (, unsubscribing, ) and letting it go out of scope.  A function like this (based on code from dowser) can be used to check the gc status after each Subscriber:

``` python
def get_counts(ros_only=True):
    # from dowser code
    gc.collect()
    typecounts = {}
    for obj in gc.get_objects():
        objtype = type(obj)
        if objtype in typecounts:
            typecounts[objtype] += 1
        else:
            typecounts[objtype] = 1
    if not ros_only:
        return typecounts
    rostypes = {k:v for k,v in typecounts.iteritems() if 'ros' in str(k)}
    print("Garbage collect counts: {}".format(rostypes))
    return rostypes

```

A complete test script can be found at this Gist (plus running a talker.py publisher):  https://gist.github.com/rethink-rlinsalata/6b0d8a5bdbe82297f522
I have also included the script and sample output at the bottom of this Issue.

**Note 1**: I originally found this when debugging a rosbridge_websocket memory leak (<-- read: use case, for fixing this bug).  I have much more data and more complex test scripts that I can provide if desired, but I wanted to provide this "simplified" test case first, because I will need to tease out the rosbridge stuff from the other tests.

**Note 2**:  Using the dowser library/tool directly, was actually the key to finding the leak and made observing the leak much, much easier.  It's simple, but I highly recommend it; all you need to add it to your python program is to add these lines near the top of your program:

``` python
import cherrypy
import dowser
cherrypy.config.update({
    'environment': 'embedded', 
    'server.socket_port': 8088,
})
cherrypy.tree.mount(dowser.Root(), '/dowser')
cherrypy.engine.start()

```

---
## Diagnosis / Towards a solution..

If you comment out the Statistics Subscriber child member (as done in the sample commits below), and run the tests again, you will see the counts go up, and then (properly) back down, without leaking.
- https://github.com/rethink-rlinsalata/ros_comm/commit/54fd2093b3da09efdd060649b0421945f29f4179
- https://github.com/rethink-rlinsalata/ros_comm/tree/fix_memleak_statistics-cut

I question why the class would be instantiated at all when statistics are disabled, rather than just bringing the enable/disable "switch" up a level, as I tried to do in this workaround, however, the enable/disable mechanism presents issues in and of itself:
- rethink-rlinsalata@5aed12984a10db17e0960ca43b97208305824926
- https://github.com/rethink-rlinsalata/ros_comm/compare/fix_memleak_statistics-upstream

I'll reserve any more subjective remarks on solving this for the comments below.

---
#### Steps:
1. Run talker, in first terminal: `rosrun test_rospy talker.py`
2. Run test script, in second terminal: `python test_rospy_sub_memleak.py`
3. Observe uncollected counts of `_SubscriberImpl` (and Poller) go up, up, and away
#### Test Script

Also at: https://gist.github.com/rethink-rlinsalata/6b0d8a5bdbe82297f522

``` python
import gc
import rospy
from std_msgs.msg import String

SUBTOPIC = '/chatter'
MSGTYPE = String

def do_sub():
    s = rospy.Subscriber(SUBTOPIC, MSGTYPE, lambda m: m)
    rospy.sleep(1.0)
    s.unregister()

def get_counts(ros_only=True):
    # from dowser code
    gc.collect()
    typecounts = {}
    for obj in gc.get_objects():
        objtype = type(obj)
        if objtype in typecounts:
            typecounts[objtype] += 1
        else:
            typecounts[objtype] = 1
    if not ros_only:
        return typecounts
    rostypes = {k:v for k,v in typecounts.iteritems() if 'ros' in str(k)}
    print("Garbage collect counts: {}".format(rostypes))
    return rostypes

if __name__ == '__main__':
    rospy.init_node('test_rospy_multisub_memory')
    i=0
    get_counts()
    while not rospy.is_shutdown():
        i += 1
        print "Running test {}..".format(i),
        do_sub()
        get_counts()

```

Output:

```
$ python test_rospy_multisub-simple.py 
Garbage collect counts: {<class 'rospy.impl.tcpros_service.Service'>: 2, <class 'rospy.impl.registration.RegManager'>: 1, <class 'rospy.topics._TopicManager'>: 1, <class 'rospy.impl.registration.RegistrationListeners'>: 1, <class 'rospy.client._Unspecified'>: 1, <class 'rospy.impl.rosout.RosOutHandler'>: 1, <class 'rospy.topics.Poller'>: 1, <class 'rospy.impl.tcpros_pubsub.TCPROSHandler'>: 1, <class 'rospy.impl.tcpros_base.TCPROSServer'>: 1, <class 'rospkg.rospack.RosPack'>: 1, <class 'rospy.topics.Publisher'>: 1, <class 'rospy.impl.tcpros_base.TCPServer'>: 1, <class 'rosgraph.xmlrpc.XmlRpcNode'>: 1, <class 'rospy.topics._PublisherImpl'>: 1, <class 'rosgraph.roslogging.RosStreamHandler'>: 1, <class 'rospy.core.NullHandler'>: 1, <class 'rospy.impl.masterslave.ROSHandler'>: 1, <class 'rospy.impl.tcpros_service.TCPService'>: 2, <class 'rospy.service.ServiceManager'>: 1}
Running test 1.. Finished
Garbage collect counts: {<class 'rospy.msproxy.MasterProxy'>: 1, <class 'rospy.impl.rosout.RosOutHandler'>: 1, <class 'rospy.impl.tcpros_pubsub.TCPROSHandler'>: 1, <class 'rospy.impl.tcpros_base.TCPROSServer'>: 1, <class 'rospy.impl.tcpros_base.TCPServer'>: 1, <class 'rospy.impl.tcpros_pubsub.TCPROSPub'>: 1, <class 'rospy.core.NullHandler'>: 1, <class 'rospy.impl.tcpros_service.Service'>: 2, <class 'rospy.topics._TopicManager'>: 1, <class 'rospy.impl.registration.RegistrationListeners'>: 1, <class 'rospy.topics.Poller'>: 2, <class 'rospy.topics._PublisherImpl'>: 1, <class 'rospy.topics._SubscriberImpl'>: 1, <class 'rospy.impl.tcpros_base.TCPROSTransport'>: 1, <class 'rosgraph.roslogging.RosStreamHandler'>: 1, <class 'rospy.impl.masterslave.ROSHandler'>: 1, <class 'rospy.impl.tcpros_service.TCPService'>: 2, <class 'rospy.client._Unspecified'>: 1, <class 'rospy.topics.Publisher'>: 1, <class 'rosgraph.xmlrpc.XmlRpcNode'>: 1, <class 'rospy.impl.tcpros_pubsub.QueuedConnection'>: 1, <class 'rospy.impl.registration.RegManager'>: 1, <class 'rospkg.rospack.RosPack'>: 1, <class 'rospy.service.ServiceManager'>: 1}
Running test 2.. Finished
Garbage collect counts: {<class 'rospy.msproxy.MasterProxy'>: 1, <class 'rospy.impl.rosout.RosOutHandler'>: 1, <class 'rospy.impl.tcpros_pubsub.TCPROSHandler'>: 1, <class 'rospy.impl.tcpros_base.TCPROSServer'>: 1, <class 'rospy.impl.tcpros_base.TCPServer'>: 1, <class 'rospy.impl.tcpros_pubsub.TCPROSPub'>: 1, <class 'rospy.core.NullHandler'>: 1, <class 'rospy.impl.tcpros_service.Service'>: 2, <class 'rospy.topics._TopicManager'>: 1, <class 'rospy.impl.registration.RegistrationListeners'>: 1, <class 'rospy.topics.Poller'>: 3, <class 'rospy.topics._PublisherImpl'>: 1, <class 'rospy.topics._SubscriberImpl'>: 2, <class 'rospy.impl.tcpros_base.TCPROSTransport'>: 1, <class 'rosgraph.roslogging.RosStreamHandler'>: 1, <class 'rospy.impl.masterslave.ROSHandler'>: 1, <class 'rospy.impl.tcpros_service.TCPService'>: 2, <class 'rospy.client._Unspecified'>: 1, <class 'rospy.topics.Publisher'>: 1, <class 'rosgraph.xmlrpc.XmlRpcNode'>: 1, <class 'rospy.impl.tcpros_pubsub.QueuedConnection'>: 1, <class 'rospy.impl.registration.RegManager'>: 1, <class 'rospkg.rospack.RosPack'>: 1, <class 'rospy.service.ServiceManager'>: 1}
^CRunning test 3.. Finished
Garbage collect counts: {<class 'rospy.topics._SubscriberImpl'>: 3, <class 'rospy.msproxy.MasterProxy'>: 1, <class 'rospy.topics._TopicManager'>: 1, <class 'rospy.impl.registration.RegistrationListeners'>: 1, <class 'rospy.client._Unspecified'>: 1, <class 'rospy.impl.rosout.RosOutHandler'>: 1, <class 'rospy.impl.tcpros_base.TCPROSServer'>: 1, <class 'rospy.impl.tcpros_pubsub.TCPROSHandler'>: 1, <class 'rospy.topics._PublisherImpl'>: 1, <class 'rospy.topics.Poller'>: 4, <class 'rospy.topics.Publisher'>: 1, <class 'rospy.impl.tcpros_base.TCPROSTransport'>: 1, <class 'rospy.impl.tcpros_base.TCPServer'>: 1, <class 'rospy.impl.tcpros_pubsub.QueuedConnection'>: 1, <class 'rosgraph.roslogging.RosStreamHandler'>: 1, <class 'rospy.core.NullHandler'>: 1, <class 'rospkg.rospack.RosPack'>: 1, <class 'rospy.service.ServiceManager'>: 1}

```

**Comment by rethink-rlinsalata** [comment]
Follow-up on the possible fix / changes referred to above, and here:
- rethink-rlinsalata@5aed12984a10db17e0960ca43b97208305824926
- https://github.com/rethink-rlinsalata/ros_comm/compare/fix_memleak_statistics-upstream
1. I believe creating (and calling) the child member object is a waste when statistics is disabled, so the enable/disable "switch" should be "moved up a level" to when the SubscriberImpl is instantiated and the member assigned.
2. However, I am not satisfied with the changes I made, and would like feedback b/c:
   1. I am not sure on the proper way to import the `get_param()` function into topics.py -- I think I'm cheating somehow currently, but I saw a similar reference without direct import elsewhere in the code, and `from rospy import get_param` gave me an error (as did `from rospy.client`).
   2. This also (maybe?) introduced a new dependency into topics?
   3. Will this increase the lag of initiailizing a new Subscriber and getting data? (it may not, I'm just not sure how big an impact it will be)
   4. I'm not sure using the parameter server is the right move in the class described as (topics.py#L448-ish): "`#  Implementation note: Subscriber attaches to a _SubscriberImpl singleton for that topic.  The underlying impl manages the connections for that publication and enables thread-safe access`" (see next note)
3. Isn't the parameter server not thread-safe in the first place? Doesn't this (and the original `SubscriberStatisticsLogger.__init__()`) make this code no longer thread-safe?

**Comment by dirk-thomas** [comment]
The `MasterProxy` API has actually been made thread-safe in _Indigo_: #459

The parameter functions are using that but currently still use a not thread-safe init logic. I created a PR to also make the initialization thread-safe without requiring locks for every access: #523

Regarding the `SubscriberStatisticsLogger` I have created #524 which:
- avoids storing a reference to the subscriber (which should remove the circular dependency and therefore the memory leak)
- moved the check if statistics are enable one level higher without moving any `knowledge` out of the `SubscriberStatisticsLogger` class (compared to your patch https://github.com/rethink-rlinsalata/ros_comm/commit/5aed12984a10db17e0960ca43b97208305824926)

Can you please check the patch from 99498b13361c995350a6182b4496f61478697e4a and confirm that it fixes the problem for you, too? If you use the complete PR please make sure to enable statistics - otherwise there won't be an instance created anymore.

**Comment by rethink-rlinsalata** [comment]
The patch from 99498b1 seems to fix the problem I saw and passes my test.  I tested both with and without statistics enabled.

However, I did notice that there is an extra `_PublisherImpl` (of /statistics) created when statistics is enabled (via the param server), that just sticks around, even though the `Publisher` is destroyed.  It is not as bad a problem as the original `_SubscriberImpl`, because it doesn't continuously increase, and is reused everytime and even after disabling statistics and turning it back on.

**Comment by dirk-thomas** [comment]
Good catch. Can you please confirm that b541ce6c640ed69cb0b217e594b304e0f0fd08dd fixes that too?

**Comment by rethink-rlinsalata** [comment]
Looks great! That fixes the publisher too.  Every thing cleans up beautifully; I can even switch enable_statistics on and off and it's fine. Thanks for your work and help on this.

**Comment by dirk-thomas** [comment]
Fixed with #524.
